Emoji panel and clipboard history: locate correct items if controls are different due to recent changes

josephsl opened the pull request:
### Link to issue number:
Fixes #10377 

### Summary of the issue:
In May 2019 Update, wrong item is announced when emoji panel and clipboard history opens due to UIA tree changes.

### Description of how this pull request fixes the issue:
Depending on UIA automation ID, select the right child from emoji panel/clipboard history panel so the correct item can be announced when it opens.

### Testing performed:
Tested via Windows 10 App Essentials add-on.

### Known issues with pull request:
None

### Change log entry:
None

josephsl committed:
appModules/modern keyboard: update copyright headers.

josephsl committed:
Modern keyboard: rename input panel object and automation ID variables. Re #10377.

For ease of readability, rename childAutomationID to inputPanelAutomationID, and assign obj.firstChild to inputPanel to better communicate event traversals.

josephsl committed:
Modern keyboard: refine object navigation so selected item for emojisand clipboard history can be announced. Re #10377.

In recent versions of emoji panel and clipboard history (particularly since may 2019 Update), something other than selected item is announced - for example, clipboard tip instead of most recently copied item, or emoji group rather than selected emoji (the latter leading to NVDA staying silent). Thus make sure to keep in mind by checking for UIA automation ID's for both old and new versions. moving to a different element if the automation ID is different from old releases.

josephsl committed:
Modenr keyboard: lint.

josephsl committed:
Emoji panel and clipboard history: locate controls instead of working with indecies. Re #10377.

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -68,35 +68,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
+		if (
+			winVersion.winVersion.build <= 17134
+			and inputPanelAutomationID in (
+				"TEMPLATE_PART_ExpressiveInputFullViewFuntionBarItemControl",
```
Could you store these in a constant?

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -68,35 +68,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
+		):
 			try:
-				self.event_UIA_elementSelected(obj.firstChild.firstChild.firstChild, nextHandler)
+				self.event_UIA_elementSelected(inputPanel.firstChild.firstChild, nextHandler)
```
I'm curious what @michaelDCurran thinks here, but I'd prefer something like `eventHandler.executeEvent` instead. IN the current code, it is difficult to find out how the flow of events is.

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -68,35 +68,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
+				emojisList = emojisList.previous
 			try:
-				self.event_UIA_elementSelected(obj.firstChild.children[-2].firstChild.firstChild, nextHandler)
+				self.event_UIA_elementSelected(emojisList.firstChild.firstChild, nextHandler)
```
Same as above

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -68,35 +68,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
+			clipboardHistory = obj.children[-2]
+			if clipboardHistory.UIAElement.cachedAutomationID == inputPanelAutomationID:
+				clipboardHistory = clipboardHistory.next
+			self.event_UIA_elementSelected(clipboardHistory, nextHandler)
```
Same as above

josephsl commented:
Hi, I’ll experiment with executeEvent in the add-on today, and as for using isWin10, that’s something I’ve considered last weekend (and will do so). As for emoji panel constants in Version 1709 and 1803, yes, I’ll do so in a list. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Monday, October 14, 2019 12:04 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Emoji panel and clipboard history: locate correct items if controls are different due to recent changes (#10378)

 

@leonardder requested changes on this pull request.

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/10378#discussion_r334344402> :

>              # Emoji panel for build 16299 and 17134.
               # This event is properly raised in build 17134.
-              if winVersion.winVersion.build <= 17134 and childAutomationID in ("TEMPLATE_PART_ExpressiveInputFullViewFuntionBarItemControl", "TEMPLATE_PART_ExpressiveInputFullViewFuntionBarCloseButton"):
+              if (
+                      winVersion.winVersion.build <= 17134

Could we use winVersion.isWin10 instead?

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/10378#discussion_r334344546> :

>              # Emoji panel for build 16299 and 17134.
               # This event is properly raised in build 17134.
-              if winVersion.winVersion.build <= 17134 and childAutomationID in ("TEMPLATE_PART_ExpressiveInputFullViewFuntionBarItemControl", "TEMPLATE_PART_ExpressiveInputFullViewFuntionBarCloseButton"):
+              if (
+                      winVersion.winVersion.build <= 17134
+                      and inputPanelAutomationID in (
+                              "TEMPLATE_PART_ExpressiveInputFullViewFuntionBarItemControl",

Could you store these in a constant?

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/10378#discussion_r334344838> :

>                      try:
-                              self.event_UIA_elementSelected(obj.firstChild.firstChild.firstChild, nextHandler)
+                              self.event_UIA_elementSelected(inputPanel.firstChild.firstChild, nextHandler)

I'm curious what @michaelDCurran <https://github.com/michaelDCurran>  thinks here, but I'd prefer something like eventHandler.executeEvent instead. IN the current code, it is difficult to find out how the flow of events is.

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/10378#discussion_r334345100> :

>                      try:
-                              self.event_UIA_elementSelected(obj.firstChild.children[-2].firstChild.firstChild, nextHandler)
+                              self.event_UIA_elementSelected(emojisList.firstChild.firstChild, nextHandler)

Same as above

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/10378#discussion_r334345210> :

>                      except AttributeError:
                               # In build 18272's emoji panel, emoji list becomes empty in some situations.
                               pass
               # Clipboard history.
               # Move to clipboard list so element selected event can pick it up.
               # #9103: if clipboard is empty, a status message is displayed instead, and luckily it is located where clipboard data items can be found.
-              elif childAutomationID == "TEMPLATE_PART_ClipboardTitleBar":
-                      self.event_UIA_elementSelected(obj.children[-2], nextHandler)
+              elif inputPanelAutomationID == "TEMPLATE_PART_ClipboardTitleBar":
+                      # Under some cases, clipboard tip text isn't shown on screen,
+                      # causing clipboard history title to be announced instead of most recently copied item.
+                      clipboardHistory = obj.children[-2]
+                      if clipboardHistory.UIAElement.cachedAutomationID == inputPanelAutomationID:
+                              clipboardHistory = clipboardHistory.next
+                      self.event_UIA_elementSelected(clipboardHistory, nextHandler)

Same as above

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/10378?email_source=notifications&email_token=AB4AXEETAF3S6GTW556AHT3QOQKWBA5CNFSM4JAFHWCKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHZNL3Y#pullrequestreview-301127151> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEDOADQ57POWP5W346DQOQKWBANCNFSM4JAFHWCA> .

josephsl committed:
Emoji panel: use winVersion.isWin10 function to determine build ranges.

josephsl committed:
Modern keyboard: call eventHandler.executeEvent rather than calling events directly. Re #10377.

Reviewed by Leonard de Ruijter: try using eventHandler.executeEvent when firing UIA item selected event for various modern keyboard features.

josephsl committed:
Emoji panel: separate 1709 and 1803 automation ID's to its own private tuple. Re #10377.

josephsl commented:
Hi,

Deferred to after 2019.3 release. Thanks.

josephsl commented:
Hi,

Now that 2019.3 stable version is out, let's get this PR going again. Note that this PR might be influenced by a UIA IME branch being worked on by @MichaelDCurran.

Thanks.

josephsl commented:
Hi,

I'm looking into an alternative implementation that may future-proof this and could serve as a test bed for File Explorer status bar announcement issue (don't remember exactly which ticket it was). basically I'm thinking about locating a specific UIA element based on automation ID rather than traversing through the UIA tree. If it works (to be tested via Windows 10 App Essentials add-on), I'll open a successor PR to resolve this unless it is fine to continue working on this branch.

Thanks.

josephsl commented:
Hi,

Actually, scratch the alternative - traversal is still required when reading clipboard announcement when clipboard history is empty. So traversal is the way to go for this one.

Thanks.

josephsl commented:
Potentially blocked by #11214. Once that PR is included in master, I suggest merging this PR, as the follow-up described in the mentioned PR will apply to this work also.

Thanks.

feerrenrut commented:
@josephsl let me know if I misunderstood please.

This PR is blocked requiring #11214 in order to be able traverse the emoji and clipboard history. Labeling as blocked.

josephsl commented:
Hi,

Technically, #11214 addresses name change event in emoji panel. This PR addresses UIA element traversal, which is different from that PR (note: the above comment was corrected).

Thanks.

josephsl commented:
Hi,

Since #11447 was merged, this pull request has become a bit complicated, as modern keyboard support must be revised extensively thanks to detection of UIA events for elements with no Automation Id property. Therefore a future PR will succeed this one - closing for now (I plan to integrate this PR into the succeeding PR).

Thanks.
